sites/funcampus/src/backend/base/cache.py

```diff
@@ -0,0 +1,184 @@
+"""
+    RedisCache with a fallback cache to prevent denial of service if Redis is down
+    Freely inspired by django-cache-fallback
+
+    Credits:
+    - https://github.com/Kub-AT/django-cache-fallback/
+"""
+
+import logging
+
+from django.conf import settings
+from django.core.cache import caches
+from django.core.cache.backends.base import BaseCache
+
+from cms.cache import CMS_PAGE_CACHE_VERSION_KEY
+from cms.utils.conf import get_cms_setting
+from django_redis.cache import RedisCache
+
+DJANGO_REDIS_LOGGER = getattr(settings, "DJANGO_REDIS_LOGGER", __name__)
+logger = logging.getLogger(DJANGO_REDIS_LOGGER)
+
+
+class RedisCacheWithFallback(BaseCache):
+    """
+    BaseCache object with a redis_cache used as main cache
+    and used the "fallback" aliased cachewhich takes over
+    in case redis_cache is down.
+    """
+
+    def __init__(self, server, params):
+        """
+        Instantiate the Redis Cache with server and params
+        and retrieve the cache with alias "fallback"
+        """
+        super().__init__(params)
+        self._redis_cache = RedisCache(server, params)
+        self._fallback_cache = caches["fallback"]
+
+    def _call_with_fallback(self, method, *args, **kwargs):
+        """
+        Try first to exec provided method through Redis cache instance,
+        in case of failure, logger reports the exception and
+        the fallback cache takes over.
+        """
+        try:
+            next_cache_state = self._call_redis_cache(method, args, kwargs)
+            self._invalidate_fallback_cache()
```

Comment:
I'm not comfortable with the fact that this function is called systematically.

One might think that it's okay because the fallback cache is in-memory and operations are fast. But the fallback cache backend is configurable and could be a remote cache, which involves more I/O expensive operations. It could reduce performances even in a healthy situation.

An improvement to solve this problem could be to store the «last successfully used cache backend». And invalidate the fallback cache only if it changes ?

Reply:
Yes, this is the big topic of this PR with @sampaccoud.

From my undestanding, it's tricky to share mutable class attribute between class instance. So we have to store this information directly in the fallback cache.

In this case, we have to get the value in the cache then delete the key if needed (2 operations). So @sampaccoud suggests to just make one operation all time : delete the CMS Page cache key.

Reply:
@madmatah makes a good point on the fact that this cache could be somewhere else (although not in this repository which is down the food chain...). So if we go back to which we want to clean this fallback cache in the first place... ie avoid going back to an old version of the site following a switch back and forth between the Redis and the fallback cache... the benefits may not outweigh the costs... The fallback cache is a degraded situation so I think we should take the risk of displaying an old version of the page. Note that the duration of the fallback cache can be set to something low like 10 minutes?

Reply:
Or clean but only on round minutes? 😅

Reply:
> From my undestanding, it's tricky to share mutable class attribute between class instance. So we have to store this information directly in the fallback cache.

My bad, I thought there was only one instance of the `RedisCacheWithFallback` class.
My suggestion is not the right way to do it, then :)

Reply:
Yes @sampaccoud, let's debounce invalidation each minutes.